In [27]:
import sys
from pathlib import Path

import pandas as pd


# Detect whether the notebook is running in Google Colab.
IN_COLAB = "google.colab" in sys.modules


# Find the repository root.
if IN_COLAB:
    from google.colab import drive  # pyright: ignore[reportMissingImports]

    drive.mount("/content/drive", force_remount=False)

    REPO_ROOT = Path(
        "/content/drive/Shareddrives/CMPE 295A_B/subset"
    )
else:
    REPO_ROOT = Path.cwd()

    while (
        REPO_ROOT.name != "masters-project"
        and REPO_ROOT.parent != REPO_ROOT
    ):
        REPO_ROOT = REPO_ROOT.parent

    if REPO_ROOT.name != "masters-project":
        raise RuntimeError(
            "Could not find repository root named 'masters-project'."
        )


# Define pipeline and data paths.
PIPELINE_DIR = REPO_ROOT / "pipelines" / "google_subset"
DATA_RAW_DIR = REPO_ROOT / "data" / "raw"
DATA_PROCESSED_DIR = (
    REPO_ROOT / "data" / "processed" / "google_subset_v1"
)


# Use the current raw-data location first.
RAW_CSV_PATH = REPO_ROOT / "subset" / "borg_traces_data.csv"

# Fall back to the standardized raw-data location if necessary.
if not RAW_CSV_PATH.exists():
    RAW_CSV_PATH = (
        DATA_RAW_DIR / "subset" / "borg_traces_data.csv"
    )

if not RAW_CSV_PATH.exists():
    raise FileNotFoundError(
        f"Raw dataset was not found at: {RAW_CSV_PATH}"
    )


# Create the processed-data directory.
DATA_PROCESSED_DIR.mkdir(
    parents=True,
    exist_ok=True
)


# Load the complete dataset.
pd.set_option("display.max_columns", None)
pd.set_option("display.width", 160)

df = pd.read_csv(
    RAW_CSV_PATH,
    low_memory=False
)


# Perform an initial dataset check.
print("Repository root:", REPO_ROOT)
print("Pipeline directory:", PIPELINE_DIR)
print("Raw dataset path:", RAW_CSV_PATH)
print("Running in Colab:", IN_COLAB)

print(
    f"\nDataset size: "
    f"{df.shape[0]:,} rows × {df.shape[1]} columns"
)

print("\nColumns:")
print(df.columns.tolist())

print("\nData types:")
display(df.dtypes)

print("\nMissing-value percentages:")
display(
    df.isna()
      .mean()
      .mul(100)
      .sort_values(ascending=False)
)

print("\nSample rows:")
display(df.head(3))

Repository root: /Users/meghanakoti/masterproject/masters-project
Pipeline directory: /Users/meghanakoti/masterproject/masters-project/pipelines/google_subset
Raw dataset path: /Users/meghanakoti/masterproject/masters-project/subset/borg_traces_data.csv
Running in Colab: False

Dataset size: 405,894 rows × 34 columns

Columns:
['Unnamed: 0', 'time', 'instance_events_type', 'collection_id', 'scheduling_class', 'collection_type', 'priority', 'alloc_collection_id', 'instance_index', 'machine_id', 'resource_request', 'constraint', 'collections_events_type', 'user', 'collection_name', 'collection_logical_name', 'start_after_collection_ids', 'vertical_scaling', 'scheduler', 'start_time', 'end_time', 'average_usage', 'maximum_usage', 'random_sample_usage', 'assigned_memory', 'page_cache_memory', 'cycles_per_instruction', 'memory_accesses_per_instruction', 'sample_rate', 'cpu_usage_distribution', 'tail_cpu_usage_distribution', 'cluster', 'event', 'failed']

Data types:


Unnamed: 0                           int64
time                                 int64
instance_events_type                 int64
collection_id                        int64
scheduling_class                     int64
collection_type                      int64
priority                             int64
alloc_collection_id                  int64
instance_index                       int64
machine_id                           int64
resource_request                    object
constraint                          object
collections_events_type              int64
user                                object
collection_name                     object
collection_logical_name             object
start_after_collection_ids          object
vertical_scaling                   float64
scheduler                          float64
start_time                           int64
end_time                             int64
average_usage                       object
maximum_usage                       object
random_samp


Missing-value percentages:


cycles_per_instruction             30.719350
memory_accesses_per_instruction    30.719350
scheduler                           0.236269
vertical_scaling                    0.236269
resource_request                    0.190690
tail_cpu_usage_distribution         0.000000
cpu_usage_distribution              0.000000
sample_rate                         0.000000
cluster                             0.000000
event                               0.000000
page_cache_memory                   0.000000
assigned_memory                     0.000000
random_sample_usage                 0.000000
maximum_usage                       0.000000
average_usage                       0.000000
end_time                            0.000000
start_time                          0.000000
Unnamed: 0                          0.000000
time                                0.000000
start_after_collection_ids          0.000000
collection_logical_name             0.000000
collection_name                     0.000000
user      


Sample rows:


,Unnamed: 0,time,instance_events_type,collection_id,scheduling_class,collection_type,priority,alloc_collection_id,instance_index,machine_id,resource_request,constraint,collections_events_type,user,collection_name,collection_logical_name,start_after_collection_ids,vertical_scaling,scheduler,start_time,end_time,average_usage,maximum_usage,random_sample_usage,assigned_memory,page_cache_memory,cycles_per_instruction,memory_accesses_per_instruction,sample_rate,cpu_usage_distribution,tail_cpu_usage_distribution,cluster,event,failed
0,0,0,2,94591244395,3,1,200,0,144,168846390496,"{'cpus': 0.020660400390625, 'memory': 0.014434...",[],2,fn8Ve4Tdl/FVVvwXFGIKe4+Wo4zLjUL/557qdFVYu5M=,Hzsv/gF8CPQXdqpsfovDTC1TJNyphDxPu7vaTeNxA74=,YCuhYrnORLiUh9WGL5q5tkBevfwtucSnFr2qPZh6Kes=,[],1.0,0.0,274800000000,275100000000,"{'cpus': 0.00466156005859375, 'memory': 0.0059...","{'cpus': 0.01190185546875, 'memory': 0.0059356...","{'cpus': 0.0043487548828125, 'memory': None}",0.014435,0.000415,NaN,NaN,1.0,[0.00314331 0.00381088 0.00401306 0.00415039 0...,[0.00535583 0.00541687 0.00548553 0.00554657 0...,7,FAIL,1
1,1,2517305308183,2,260697606809,2,0,360,221495397286,335,85515092,"{'cpus': 0.00724029541015625, 'memory': 0.0013...",[],2,DrrEIEWkWuW7RrZwpHLCN0k0A2J0usJeyt3wtqzZ7Kk=,hDGffcrF/rhQQEG8Uns/RMUK7R15DXjFnRasoKFhefI=,wcRcAMuop2OqH9EW4feH919tadFec5a11ply0hcS/C8=,[],2.0,0.0,1800713000000,1800714000000,"{'cpus': 0.0, 'memory': 9.5367431640625e-07}","{'cpus': 0.0, 'memory': 9.5367431640625e-07}","{'cpus': 0.0, 'memory': None}",0.000000,0.000000,NaN,NaN,1.0,[1.23977661e-05 1.23977661e-05 1.23977661e-05 ...,[1.23977661e-05 1.23977661e-05 1.23977661e-05 ...,7,FAIL,1
2,2,195684022913,6,276227177776,2,0,103,0,376,169321752432,"{'cpus': 0.048583984375, 'memory': 0.004165649...",[],6,/ivQBmewiFcXfGJdCUsEKx47NiRE29Tjiq3gw+zR2Cg=,kk6+maA6fvAdJ+VTU8AcpzQPTyVrx+ySt0MXRAyO8FU=,zCA2dl2PDptd82Hob906gE82JHzx0SbqA4mZurqZdmY=,[],2.0,1.0,81300000000,81600000000,"{'cpus': 0.024200439453125, 'memory': 0.002788...","{'cpus': 0.06005859375, 'memory': 0.0028457641...","{'cpus': 0.026458740234375, 'memory': None}",0.010422,0.000235,0.939919,0.001318,1.0,[0.01344299 0.01809692 0.0201416 0.02246094 0...,[0.02902222 0.02929688 0.0295105 0.0296936 0...,7,SCHEDULE,0


## Extract resource request and usage values

The raw CSV stores resource values as dictionary-like strings.
This step safely parses those fields and creates numeric columns for
requested, average, and peak CPU and memory.

Missing values are preserved, and no rows or events are filtered.
Units are not converted because they are handled in a later validation step.

In [28]:
# =========================================================
# step 2: Parse nested resource fields
# =========================================================

import ast


# These columns contain dictionary-like strings.
NESTED_COLUMNS = [
    "resource_request",
    "average_usage",
    "maximum_usage",
    "random_sample_usage",
]

missing_columns = [
    column for column in NESTED_COLUMNS
    if column not in df.columns
]

if missing_columns:
    raise ValueError(
        f"Missing required nested columns: {missing_columns}"
    )


# Convert a dictionary-like value into a Python dictionary.
# Invalid or missing values become None.
def parse_nested(value):
    if value is None:
        return None

    if isinstance(value, float) and pd.isna(value):
        return None

    if isinstance(value, dict):
        return value

    if isinstance(value, str):
        try:
            parsed = ast.literal_eval(value)

            if isinstance(parsed, dict):
                return parsed

        except (ValueError, SyntaxError):
            pass

    return None


# Read one field from a parsed dictionary.
def get_nested_value(value, key):
    if isinstance(value, dict):
        return value.get(key)

    return None


# Parse nested columns and record missing values.
parsed_nested = {}
audit_rows = []

for column in NESTED_COLUMNS:
    parsed = df[column].map(parse_nested)
    parsed_nested[column] = parsed

    cpu_values = parsed.map(
        lambda value: get_nested_value(value, "cpus")
    )

    memory_values = parsed.map(
        lambda value: get_nested_value(value, "memory")
    )

    audit_rows.append(
        {
            "source_column": column,
            "missing_or_invalid_dict": int(parsed.isna().sum()),
            "missing_cpu_value": int(cpu_values.isna().sum()),
            "missing_memory_value": int(memory_values.isna().sum()),
        }
    )


# Store the parsing audit for review.
nested_audit = pd.DataFrame(audit_rows)

print("Nested resource-field audit:")
display(nested_audit)


# Extract numeric raw resource values.
# The raw_ prefix indicates that units have not been converted.
df["raw_req_cpu"] = pd.to_numeric(
    parsed_nested["resource_request"].map(
        lambda value: get_nested_value(value, "cpus")
    ),
    errors="coerce",
)

df["raw_req_memory"] = pd.to_numeric(
    parsed_nested["resource_request"].map(
        lambda value: get_nested_value(value, "memory")
    ),
    errors="coerce",
)

df["raw_avg_cpu"] = pd.to_numeric(
    parsed_nested["average_usage"].map(
        lambda value: get_nested_value(value, "cpus")
    ),
    errors="coerce",
)

df["raw_avg_memory"] = pd.to_numeric(
    parsed_nested["average_usage"].map(
        lambda value: get_nested_value(value, "memory")
    ),
    errors="coerce",
)

df["raw_peak_cpu"] = pd.to_numeric(
    parsed_nested["maximum_usage"].map(
        lambda value: get_nested_value(value, "cpus")
    ),
    errors="coerce",
)

df["raw_peak_memory"] = pd.to_numeric(
    parsed_nested["maximum_usage"].map(
        lambda value: get_nested_value(value, "memory")
    ),
    errors="coerce",
)


# Check the new numeric columns.
RAW_RESOURCE_COLUMNS = [
    "raw_req_cpu",
    "raw_req_memory",
    "raw_avg_cpu",
    "raw_avg_memory",
    "raw_peak_cpu",
    "raw_peak_memory",
]

print("\nExtracted raw resource values:")
display(df[RAW_RESOURCE_COLUMNS].head())

print("\nMissing percentages after extraction:")
display(
    df[RAW_RESOURCE_COLUMNS]
      .isna()
      .mean()
      .mul(100)
      .sort_values(ascending=False)
)

Nested resource-field audit:


,source_column,missing_or_invalid_dict,missing_cpu_value,missing_memory_value
0,resource_request,774,774,774
1,average_usage,0,0,0
2,maximum_usage,0,0,4927
3,random_sample_usage,0,0,405894



Extracted raw resource values:


,raw_req_cpu,raw_req_memory,raw_avg_cpu,raw_avg_memory,raw_peak_cpu,raw_peak_memory
0,0.020660,0.014435,0.004662,5.920410e-03,0.011902,5.935669e-03
1,0.007240,0.001303,0.000000,9.536743e-07,0.000000,9.536743e-07
2,0.048584,0.004166,0.024200,2.788544e-03,0.060059,2.845764e-03
3,0.070435,0.041626,0.047607,3.442383e-02,0.133301,3.466797e-02
4,0.002449,0.000232,0.000271,7.629395e-05,0.000415,7.629395e-05



Missing percentages after extraction:


raw_peak_memory    1.213864
raw_req_cpu        0.190690
raw_req_memory     0.190690
raw_avg_cpu        0.000000
raw_avg_memory     0.000000
raw_peak_cpu       0.000000
dtype: float64

In [29]:
# =========================================================
# Step 3: Basic resource and timestamp validation
# =========================================================
#
# This cell checks that extracted values are valid.
# It does not filter rows, convert units, or create executions.
# =========================================================

RAW_RESOURCE_COLUMNS = [
    "raw_req_cpu",
    "raw_req_memory",
    "raw_avg_cpu",
    "raw_avg_memory",
    "raw_peak_cpu",
    "raw_peak_memory",
]


# Check for negative resource values.
negative_counts = (
    df[RAW_RESOURCE_COLUMNS]
      .lt(0)
      .sum()
)

if negative_counts.any():
    raise ValueError(
        f"Negative resource values found:\n{negative_counts}"
    )


# Check that peak usage is not below average usage.
cpu_check = df[
    ["raw_avg_cpu", "raw_peak_cpu"]
].dropna()

memory_check = df[
    ["raw_avg_memory", "raw_peak_memory"]
].dropna()

if (
    cpu_check["raw_peak_cpu"]
    < cpu_check["raw_avg_cpu"]
).any():
    raise ValueError(
        "Some rows have peak CPU below average CPU."
    )

if (
    memory_check["raw_peak_memory"]
    < memory_check["raw_avg_memory"]
).any():
    raise ValueError(
        "Some rows have peak memory below average memory."
    )


# Check timestamp ordering without storing a duration yet.
raw_usage_window_delta = (
    pd.to_numeric(df["end_time"], errors="coerce")
    - pd.to_numeric(df["start_time"], errors="coerce")
)

if (raw_usage_window_delta < 0).any():
    raise ValueError(
        "Some rows have negative usage-window durations."
    )

if (raw_usage_window_delta == 0).any():
    raise ValueError(
        "Some rows have zero usage-window durations."
    )


print("Basic resource and timestamp validation passed.")

Basic resource and timestamp validation passed.


In [ ]:
# =========================================================
# Cell 4: Create canonical timestamp columns
# =========================================================
#
# Google trace timestamps are stored in microseconds.
# This cell creates explicit event-time and usage-window fields.
#
# The original raw timestamp columns are preserved.
# The usage-window duration is not execution runtime.
# =========================================================

INT64_MAX = 2**63 - 1


# Convert raw timestamp columns to numeric values.
event_time_us = pd.to_numeric(
    df["time"],
    errors="coerce"
)

usage_start_time_us = pd.to_numeric(
    df["start_time"],
    errors="coerce"
)

usage_end_time_us = pd.to_numeric(
    df["end_time"],
    errors="coerce"
)


# Treat the Int64 maximum value as an invalid event timestamp.
event_time_us = (
    event_time_us
    .mask(event_time_us == INT64_MAX)
    .astype("Int64")
)

usage_start_time_us = usage_start_time_us.astype("Int64")
usage_end_time_us = usage_end_time_us.astype("Int64")


# Create explicit canonical timestamp columns.
df["event_time_us"] = event_time_us
df["usage_start_time_us"] = usage_start_time_us
df["usage_end_time_us"] = usage_end_time_us


# Calculate the resource usage-window duration in seconds.
# This is not the final reconstructed execution runtime.
df["usage_window_duration_sec"] = (
    df["usage_end_time_us"]
    - df["usage_start_time_us"]
) / 1_000_000


# Validate usage-window timestamps.
if df["usage_start_time_us"].isna().any():
    raise ValueError(
        "Missing usage start timestamps found."
    )

if df["usage_end_time_us"].isna().any():
    raise ValueError(
        "Missing usage end timestamps found."
    )

if (
    df["usage_end_time_us"]
    < df["usage_start_time_us"]
).any():
    raise ValueError(
        "Negative usage-window durations found."
    )


print("Canonical timestamp columns created.")
print(
    "Invalid event timestamps converted to missing:",
    int(df["event_time_us"].isna().sum())
)

display(
    df[
        [
            "time",
            "start_time",
            "end_time",
            "event_time_us",
            "usage_start_time_us",
            "usage_end_time_us",
            "usage_window_duration_sec",
        ]
    ].head()
)

Raw timestamp ranges:
start_time: 300,000,000 --> 2,678,400,000,000
end_time  : 600,000,000 --> 2,678,700,000,000

Candidate interpretations of the start-time span:
 nanoseconds: 0.0310 days
microseconds: 30.9965 days
milliseconds: 30,996.5278 days
     seconds: 30,996,527.7778 days

Most common raw durations:


,raw_duration,row_count
0,300000000,261434
1,1000000,30071
2,2000000,22953
3,3000000,20906
4,4000000,9058
5,241000000,6900
6,97000000,3443
7,295000000,3213
8,72000000,2160
9,41000000,2130



Raw duration converted under candidate units:
 nanoseconds: minimum = 0.001, median = 0.3, maximum = 0.3
microseconds: minimum = 1, median = 300, maximum = 300
milliseconds: minimum = 1000, median = 300000, maximum = 300000
     seconds: minimum = 1e+06, median = 3e+08, maximum = 3e+08


In [9]:
# =========================================================
# Cell 5: Create canonical time fields
# =========================================================
# - start_time and end_time are in microseconds.
# - The trace spans approximately one month.
# - Duration is converted to seconds.
#
# The original raw timestamp columns are preserved.
# Rows are not filtered in this cell.
# =========================================================

TIMESTAMP_UNIT = "microseconds"
MICROSECONDS_PER_SECOND = 1_000_000

# Preserve raw timestamp values and ensure numeric representation.
df["start_time_us"] = pd.to_numeric(
    df["start_time"],
    errors="coerce"
)

df["end_time_us"] = pd.to_numeric(
    df["end_time"],
    errors="coerce"
)

# Calculate duration in seconds using the verified unit.
df["duration_sec"] = (
    df["end_time_us"] - df["start_time_us"]
) / MICROSECONDS_PER_SECOND

# Validate timestamp conversion.
invalid_timestamp_count = int(
    df[["start_time_us", "end_time_us"]]
    .isna()
    .any(axis=1)
    .sum()
)

negative_duration_count = int(
    (df["duration_sec"] < 0).sum()
)

zero_duration_count = int(
    (df["duration_sec"] == 0).sum()
)

print("Timestamp unit:", TIMESTAMP_UNIT)
print("Invalid timestamp rows:", invalid_timestamp_count)
print("Negative-duration rows:", negative_duration_count)
print("Zero-duration rows:", zero_duration_count)

print("\nDuration in seconds:")
display(
    df["duration_sec"].describe(
        percentiles=[0.01, 0.50, 0.95, 0.99]
    )
)

print("\nCanonical time fields:")
display(
    df[
        [
            "start_time",
            "end_time",
            "start_time_us",
            "end_time_us",
            "duration_sec",
        ]
    ].head()
)

Timestamp unit: microseconds
Invalid timestamp rows: 0
Negative-duration rows: 0
Zero-duration rows: 0

Duration in seconds:


count    405894.000000
mean        211.874270
std         129.227356
min           1.000000
1%            1.000000
50%         300.000000
95%         300.000000
99%         300.000000
max         300.000000
Name: duration_sec, dtype: float64


Canonical time fields:


,start_time,end_time,start_time_us,end_time_us,duration_sec
0,274800000000,275100000000,274800000000,275100000000,300.0
1,1800713000000,1800714000000,1800713000000,1800714000000,1.0
2,81300000000,81600000000,81300000000,81600000000,300.0
3,1075500000000,1075800000000,1075500000000,1075800000000,300.0
4,1565315000000,1565317000000,1565315000000,1565317000000,2.0


In [10]:
# =========================================================
# Cell 6: Audit CPU and memory value scales
# =========================================================
#
# No unit conversion is performed in this cell.
# The values remain in their original raw scale.
# =========================================================

RESOURCE_SCALE_COLUMNS = [
    "raw_req_cpu",
    "raw_avg_cpu",
    "raw_peak_cpu",
    "raw_req_memory",
    "raw_avg_memory",
    "raw_peak_memory",
    "assigned_memory",
    "page_cache_memory",
]

# Display resource distributions.
print("Resource value summary:")

display(
    df[RESOURCE_SCALE_COLUMNS]
    .describe(
        percentiles=[
            0.01,
            0.25,
            0.50,
            0.75,
            0.95,
            0.99,
            0.999,
        ]
    )
    .T
)


# Count values greater than one.
# This is only a diagnostic; values above one are not
# automatically incorrect.
print("\nValues greater than 1:")

display(
    df[RESOURCE_SCALE_COLUMNS]
    .gt(1)
    .sum()
    .sort_values(ascending=False)
)


# Verify that peak CPU is not lower than average CPU.
cpu_check = df[
    ["raw_avg_cpu", "raw_peak_cpu"]
].dropna()

print("\nCPU relationship:")
print("Rows compared:", len(cpu_check))
print(
    "Peak CPU >= average CPU:",
    int(
        (
            cpu_check["raw_peak_cpu"]
            >= cpu_check["raw_avg_cpu"]
        ).sum()
    )
)
print(
    "Peak CPU < average CPU:",
    int(
        (
            cpu_check["raw_peak_cpu"]
            < cpu_check["raw_avg_cpu"]
        ).sum()
    )
)


# Verify that peak memory is not lower than average memory.
memory_check = df[
    ["raw_avg_memory", "raw_peak_memory"]
].dropna()

print("\nMemory relationship:")
print("Rows compared:", len(memory_check))
print(
    "Peak memory >= average memory:",
    int(
        (
            memory_check["raw_peak_memory"]
            >= memory_check["raw_avg_memory"]
        ).sum()
    )
)
print(
    "Peak memory < average memory:",
    int(
        (
            memory_check["raw_peak_memory"]
            < memory_check["raw_avg_memory"]
        ).sum()
    )
)


# Compare request and peak values.
# These are preliminary diagnostics, not final violations.
cpu_request_check = df[
    ["raw_req_cpu", "raw_peak_cpu"]
].dropna()

memory_request_check = df[
    ["raw_req_memory", "raw_peak_memory"]
].dropna()

print("\nRequest versus peak:")
print(
    "CPU rows where peak > request:",
    int(
        (
            cpu_request_check["raw_peak_cpu"]
            > cpu_request_check["raw_req_cpu"]
        ).sum()
    )
)

print(
    "Memory rows where peak > request:",
    int(
        (
            memory_request_check["raw_peak_memory"]
            > memory_request_check["raw_req_memory"]
        ).sum()
    )
)

Resource value summary:


,count,mean,std,min,1%,25%,50%,75%,95%,99%,99.9%,max
raw_req_cpu,405120.0,0.015336,0.028646,0.0,0.00000,0.004051,0.008102,0.015900,0.052856,0.125488,0.284180,0.583008
raw_avg_cpu,405894.0,0.007446,0.018567,0.0,0.00000,0.000200,0.001049,0.007286,0.032227,0.079468,0.232910,0.538086
raw_peak_cpu,405894.0,0.025352,0.052681,0.0,0.00000,0.000794,0.005005,0.029724,0.106079,0.237793,0.543945,1.271484
raw_req_memory,405120.0,0.009042,0.022300,0.0,0.00018,0.001303,0.003254,0.006508,0.036987,0.104126,0.189697,0.286621
raw_avg_memory,405894.0,0.005640,0.016559,0.0,0.00000,0.000238,0.001282,0.004257,0.019714,0.129639,0.133301,0.223877
raw_peak_memory,400967.0,0.005946,0.016632,0.0,0.00000,0.000262,0.001444,0.004776,0.020935,0.129639,0.129639,0.224365
assigned_memory,405894.0,0.008927,0.024937,0.0,0.00000,0.000196,0.002693,0.006683,0.034607,0.187500,0.199707,0.286133
page_cache_memory,405894.0,0.000398,0.000550,0.0,0.00000,0.000008,0.000227,0.000539,0.001688,0.002480,0.002869,0.008453



Values greater than 1:


raw_peak_cpu         183
raw_req_cpu            0
raw_avg_cpu            0
raw_req_memory         0
raw_avg_memory         0
raw_peak_memory        0
assigned_memory        0
page_cache_memory      0
dtype: int64


CPU relationship:
Rows compared: 405894
Peak CPU >= average CPU: 405894
Peak CPU < average CPU: 0

Memory relationship:
Rows compared: 400967
Peak memory >= average memory: 400967
Peak memory < average memory: 0

Request versus peak:
CPU rows where peak > request: 203459
Memory rows where peak > request: 73310


In [11]:
# =========================================================
# Cell 7: Compare request memory with assigned memory
# =========================================================
#
# This cell checks whether:
# - resource_request["memory"]
# - assigned_memory
#
# represent the same quantity or different memory concepts.
#
# No unit conversion or filtering is performed.
# =========================================================

memory_comparison = df[
    [
        "raw_req_memory",
        "assigned_memory",
        "page_cache_memory",
        "raw_peak_memory",
    ]
].dropna(
    subset=["raw_req_memory", "assigned_memory"]
).copy()

memory_comparison["request_minus_assigned"] = (
    memory_comparison["raw_req_memory"]
    - memory_comparison["assigned_memory"]
)

memory_comparison["request_assigned_ratio"] = np.where(
    memory_comparison["assigned_memory"] > 0,
    memory_comparison["raw_req_memory"]
    / memory_comparison["assigned_memory"],
    np.nan,
)

print("Memory comparison summary:")
display(
    memory_comparison[
        [
            "raw_req_memory",
            "assigned_memory",
            "page_cache_memory",
            "raw_peak_memory",
            "request_minus_assigned",
            "request_assigned_ratio",
        ]
    ].describe(
        percentiles=[
            0.01,
            0.25,
            0.50,
            0.75,
            0.95,
            0.99,
        ]
    ).T
)


# Check how closely request memory matches assigned memory.
print("\nRequest memory equals assigned memory:")
print(
    int(
        np.isclose(
            memory_comparison["raw_req_memory"],
            memory_comparison["assigned_memory"],
            rtol=1e-9,
            atol=1e-12,
        ).sum()
    )
)

print(
    "\nRequest memory differs from assigned memory:"
)
print(
    int(
        (
            ~np.isclose(
                memory_comparison["raw_req_memory"],
                memory_comparison["assigned_memory"],
                rtol=1e-9,
                atol=1e-12,
            )
        ).sum()
    )
)

print("\nSample memory comparisons:")
display(memory_comparison.head(10))

Memory comparison summary:


,count,mean,std,min,1%,25%,50%,75%,95%,99%,max
raw_req_memory,405120.0,0.009042,0.022300,0.000000,0.000180,0.001303,3.253937e-03,0.006508,0.036987,0.104126,0.286621
assigned_memory,405120.0,0.008913,0.024939,0.000000,0.000000,0.000196,2.658844e-03,0.006683,0.034607,0.187500,0.286133
page_cache_memory,405120.0,0.000399,0.000550,0.000000,0.000000,0.000008,2.269745e-04,0.000539,0.001692,0.002480,0.008453
raw_peak_memory,400369.0,0.005936,0.016627,0.000000,0.000000,0.000257,1.443863e-03,0.004776,0.020935,0.129639,0.224365
request_minus_assigned,405120.0,0.000129,0.014772,-0.272322,-0.083374,0.000000,5.030698e-17,0.001303,0.004196,0.055847,0.286351
request_assigned_ratio,307665.0,1.183431,3.430368,0.000000,0.173119,0.949791,1.000000e+00,1.000000,1.371824,7.556080,1061.992933



Request memory equals assigned memory:
152037

Request memory differs from assigned memory:
253083

Sample memory comparisons:


,raw_req_memory,assigned_memory,page_cache_memory,raw_peak_memory,request_minus_assigned,request_assigned_ratio
0,0.014435,0.014435,4.148483e-04,5.935669e-03,0.000000e+00,1.000000
1,0.001303,0.000000,0.000000e+00,9.536743e-07,1.302719e-03,NaN
2,0.004166,0.010422,2.346039e-04,2.845764e-03,-6.256104e-03,0.399707
3,0.041626,0.041626,2.250671e-04,3.466797e-02,0.000000e+00,1.000000
4,0.000232,0.000272,9.536743e-06,7.629395e-05,-4.005432e-05,0.852632
5,0.005409,0.005409,1.844406e-03,4.051208e-03,5.030698e-17,1.000000
6,0.001562,0.002190,5.722046e-06,1.653671e-03,-6.275177e-04,0.713415
7,0.001501,0.001474,5.912781e-05,7.820129e-04,2.670288e-05,1.018111
8,0.003613,0.006683,9.536743e-07,6.790161e-04,-3.070831e-03,0.540525
9,0.006767,0.006790,1.583099e-04,4.241943e-03,-2.288818e-05,0.996629


In [ ]:
# =========================================================
# Cell 8 — Understand event types

# to understand whether rows marked FAIL, failed tasks or different event types should be included
#
# =========================================================

EVENT_COLUMNS = [
    "event",
    "failed",
    "instance_events_type",
    "collections_events_type",
]

for column in EVENT_COLUMNS:
    print(f"\n--- {column} ---")
    display(
        df[column]
        .value_counts(dropna=False)
        .head(30)
    )


print("\nEvent versus failed cross-tabulation:")
display(
    pd.crosstab(
        df["event"],
        df["failed"],
        dropna=False
    )
)


print("\nFailed-value percentages:")
display(
    df["failed"]
    .value_counts(normalize=True, dropna=False)
    .mul(100)
    .rename("percentage")
)


print("\nFailed rows by event:")
display(
    pd.crosstab(
        df["event"],
        df["failed"],
        normalize="index",
        dropna=False
    ).mul(100)
)


--- event ---


event
FINISH            92867
FAIL              92678
ENABLE            75907
SCHEDULE          69104
LOST              59515
EVICT             14756
KILL                951
UPDATE_PENDING      111
QUEUE                 4
UPDATE_RUNNING        1
Name: count, dtype: int64


--- failed ---


failed
0    313216
1     92678
Name: count, dtype: int64


--- instance_events_type ---


instance_events_type
3     92867
2     92678
0     75907
6     69104
5     59515
1     14756
4       951
9       111
8         4
10        1
Name: count, dtype: int64


--- collections_events_type ---


collections_events_type
3     92867
2     92678
0     75907
6     69104
5     59515
1     14756
4       951
9       111
8         4
10        1
Name: count, dtype: int64


Event versus failed cross-tabulation:


failed,0,1
event,,
ENABLE,75907,0
EVICT,14756,0
FAIL,0,92678
FINISH,92867,0
KILL,951,0
LOST,59515,0
QUEUE,4,0
SCHEDULE,69104,0
UPDATE_PENDING,111,0



Failed-value percentages:


failed
0    77.166945
1    22.833055
Name: percentage, dtype: float64


Failed rows by event:


failed,0,1
event,,
ENABLE,100.0,0.0
EVICT,100.0,0.0
FAIL,0.0,100.0
FINISH,100.0,0.0
KILL,100.0,0.0
LOST,100.0,0.0
QUEUE,100.0,0.0
SCHEDULE,100.0,0.0
UPDATE_PENDING,100.0,0.0


In [14]:
# =========================================================
# Cell 9: Compare resource completeness by event type
# =========================================================
#Cell 9 — Check data quality by event. This cell checks whether request, usage, and duration values are available for each event type.These cells help us decide later which event records
#should be included. They do not modify or filter the data.
#
#
# =========================================================

event_summary = (
    df.groupby("event", dropna=False)
    .agg(
        row_count=("event", "size"),
        failed_count=("failed", "sum"),
        missing_req_cpu=("raw_req_cpu", lambda x: x.isna().sum()),
        missing_req_memory=(
            "raw_req_memory",
            lambda x: x.isna().sum(),
        ),
        missing_peak_cpu=("raw_peak_cpu", lambda x: x.isna().sum()),
        missing_peak_memory=(
            "raw_peak_memory",
            lambda x: x.isna().sum(),
        ),
        min_duration_sec=("duration_sec", "min"),
        median_duration_sec=("duration_sec", "median"),
        max_duration_sec=("duration_sec", "max"),
    )
    .sort_values("row_count", ascending=False)
)

print("Resource completeness and duration by event:")
display(event_summary)


# Show event shares for easier comparison.
event_summary["row_percentage"] = (
    event_summary["row_count"]
    / len(df)
    * 100
)

print("\nEvent percentages:")
display(
    event_summary[
        ["row_count", "row_percentage", "failed_count"]
    ]
)

Resource completeness and duration by event:


,row_count,failed_count,missing_req_cpu,missing_req_memory,missing_peak_cpu,missing_peak_memory,min_duration_sec,median_duration_sec,max_duration_sec
event,,,,,,,,,
FINISH,92867,0,0,0,0,47,1.0,300.0,300.0
FAIL,92678,92678,361,361,0,913,1.0,300.0,300.0
ENABLE,75907,0,413,413,0,498,1.0,300.0,300.0
SCHEDULE,69104,0,0,0,0,3469,1.0,300.0,300.0
LOST,59515,0,0,0,0,0,1.0,300.0,300.0
EVICT,14756,0,0,0,0,0,1.0,4.0,300.0
KILL,951,0,0,0,0,0,1.0,117.0,300.0
UPDATE_PENDING,111,0,0,0,0,0,2.0,110.0,300.0
QUEUE,4,0,0,0,0,0,21.0,52.0,83.0



Event percentages:


,row_count,row_percentage,failed_count
event,,,
FINISH,92867,22.879619,0
FAIL,92678,22.833055,92678
ENABLE,75907,18.701188,0
SCHEDULE,69104,17.025135,0
LOST,59515,14.662695,0
EVICT,14756,3.635432,0
KILL,951,0.234298,0
UPDATE_PENDING,111,0.027347,0
QUEUE,4,0.000985,0


In [15]:
# =========================================================
# Cell 10: Examine candidate execution structure
# =========================================================

# Create a temporary identifier for diagnostic analysis.
df["_candidate_execution_id"] = (
    df["collection_id"].astype(str)
    + "_"
    + df["instance_index"].astype(str)
)


# Summarize raw event rows for each candidate execution.
candidate_execution_summary = (
    df.groupby("_candidate_execution_id")
    .agg(
        raw_row_count=("event", "size"),
        event_type_count=("event", "nunique"),
        event_types=(
            "event",
            lambda values: tuple(sorted(values.unique()))
        ),
        failed_row_count=("failed", "sum"),
        start_time_us=("start_time_us", "min"),
        end_time_us=("end_time_us", "max"),
    )
)


print("Candidate execution summary:")
display(candidate_execution_summary.head(10))


# Show the distribution of raw rows per candidate execution.
print("\nRaw rows per candidate execution:")
display(
    candidate_execution_summary["raw_row_count"]
    .describe(
        percentiles=[
            0.50,
            0.75,
            0.90,
            0.95,
            0.99,
        ]
    )
)


# Show how many event types appear per candidate execution.
print("\nEvent types per candidate execution:")
display(
    candidate_execution_summary["event_type_count"]
    .value_counts()
    .sort_index()
)


# Show the most event-heavy candidate executions.
print("\nExamples with multiple event rows:")
display(
    candidate_execution_summary[
        candidate_execution_summary["raw_row_count"] > 1
    ]
    .sort_values(
        "raw_row_count",
        ascending=False
    )
    .head(10)
)

Candidate execution summary:


,raw_row_count,event_type_count,event_types,failed_row_count,start_time_us,end_time_us
_candidate_execution_id,,,,,,
100286199_134,1,1,"(ENABLE,)",0,2292000000000,2292300000000
100286199_250,1,1,"(ENABLE,)",0,2292000000000,2292300000000
100286199_269,1,1,"(ENABLE,)",0,2292000000000,2292300000000
100286199_321,1,1,"(ENABLE,)",0,2292000000000,2292300000000
100286199_323,1,1,"(ENABLE,)",0,2292000000000,2292300000000
100287787_122,1,1,"(FAIL,)",1,2248500000000,2248800000000
100288420_58,1,1,"(FINISH,)",0,976200000000,976500000000
10084165805_105,1,1,"(FAIL,)",1,958800000000,959100000000
10084165805_112,1,1,"(FAIL,)",1,958800000000,959100000000



Raw rows per candidate execution:


count    242946.000000
mean          1.670717
std          17.788001
min           1.000000
50%           1.000000
75%           1.000000
90%           2.000000
95%           3.000000
99%          10.000000
max        7704.000000
Name: raw_row_count, dtype: float64


Event types per candidate execution:


event_type_count
1    232646
2      9960
3       332
4         8
Name: count, dtype: int64


Examples with multiple event rows:


,raw_row_count,event_type_count,event_types,failed_row_count,start_time_us,end_time_us
_candidate_execution_id,,,,,,
116646838687_17,7704,2,"(ENABLE, FAIL)",1138,361200000000,2288923000000
14423004927_969,1430,1,"(FINISH,)",0,1979700000000,1980000000000
122697038507_1191,1308,1,"(ENABLE,)",0,1944000000000,1944300000000
348020181306_5,1135,1,"(FAIL,)",1135,2544777000000,2544778000000
14423004927_1021,994,1,"(FINISH,)",0,1979700000000,1980000000000
260725669813_176,905,3,"(ENABLE, FINISH, LOST)",0,25800000000,1175400000000
260725669813_112,767,3,"(ENABLE, FINISH, LOST)",0,25800000000,1175400000000
122697038507_5012,686,1,"(ENABLE,)",0,1944000000000,1944300000000
128073966478_2432,587,2,"(FAIL, FINISH)",1,115800000000,430500000000


## Candidate Execution Identity Check

This analysis tested whether `collection_id + instance_index` uniquely identifies one workload execution.

The dataset contains:

- 405,894 raw rows;
- 242,946 candidate execution groups;
- median of 1 raw row per candidate group;
- maximum of 7,704 raw rows for one candidate group.

Some candidate groups contain many repeated rows and multiple event types across a large time range. For example, one candidate group contained 7,704 rows with repeated `ENABLE` and `FAIL` events.

### Finding

`collection_id + instance_index` cannot yet be treated as the final `execution_id`.

It may represent a recurring task or instance key rather than one individual execution. We must inspect repeated groups and determine how separate execution episodes should be identified before aggregating CPU, memory, duration, or targets.

This decision affects:

- execution reconstruction;
- peak CPU and memory aggregation;
- recurrence grouping;
- historical profiles;
- target construction;
- baseline evaluation.

In [16]:
# =========================================================
# Cell 11: Inspect repeated candidate execution groups
# =========================================================
#
# This cell investigates candidate IDs with many raw rows.
# It helps determine whether the candidate key represents
# one execution or multiple repeated episodes.
#
# No rows are filtered or modified.
# =========================================================

# Select the candidate groups with the most raw rows.
largest_candidate_ids = (
    candidate_execution_summary["raw_row_count"]
    .sort_values(ascending=False)
    .head(5)
    .index
)

for candidate_id in largest_candidate_ids:
    print("\n" + "=" * 80)
    print("Candidate execution:", candidate_id)

    candidate_rows = (
        df[df["_candidate_execution_id"] == candidate_id]
        .sort_values(
            ["start_time_us", "end_time_us", "time"],
            kind="mergesort"
        )
    )

    print("Raw rows:", len(candidate_rows))
    print("Event types:", candidate_rows["event"].unique())
    print("Machines:", candidate_rows["machine_id"].nunique())
    print("Distinct starts:", candidate_rows["start_time_us"].nunique())
    print("Distinct ends:", candidate_rows["end_time_us"].nunique())

    display(
        candidate_rows[
            [
                "_candidate_execution_id",
                "collection_id",
                "instance_index",
                "machine_id",
                "event",
                "failed",
                "time",
                "start_time_us",
                "end_time_us",
                "duration_sec",
                "raw_req_cpu",
                "raw_req_memory",
                "raw_peak_cpu",
                "raw_peak_memory",
            ]
        ].head(30)
    )


Candidate execution: 116646838687_17
Raw rows: 7704
Event types: ['FAIL' 'ENABLE']
Machines: 6640
Distinct starts: 2
Distinct ends: 2


,_candidate_execution_id,collection_id,instance_index,machine_id,event,failed,time,start_time_us,end_time_us,duration_sec,raw_req_cpu,raw_req_memory,raw_peak_cpu,raw_peak_memory
372523,116646838687_17,116646838687,17,10215930,FAIL,1,0,361200000000,361217000000,17.0,0.001137,0.006508,0.005554,0.000759
298051,116646838687_17,116646838687,17,66621091,FAIL,1,5273210407,361200000000,361217000000,17.0,0.001137,0.006508,0.005554,0.000759
65501,116646838687_17,116646838687,17,350633273261,FAIL,1,6806156764,361200000000,361217000000,17.0,0.001137,0.006508,0.005554,0.000759
104108,116646838687_17,116646838687,17,66621498,FAIL,1,12490474862,361200000000,361217000000,17.0,0.001137,0.006508,0.005554,0.000759
285735,116646838687_17,116646838687,17,56419634,FAIL,1,15190951363,361200000000,361217000000,17.0,0.001137,0.006508,0.005554,0.000759
78332,116646838687_17,116646838687,17,291171525,FAIL,1,18387193235,361200000000,361217000000,17.0,0.001137,0.006508,0.005554,0.000759
314415,116646838687_17,116646838687,17,741431572,FAIL,1,20063881043,361200000000,361217000000,17.0,0.001137,0.006508,0.005554,0.000759
185753,116646838687_17,116646838687,17,1702792162,FAIL,1,27208478678,361200000000,361217000000,17.0,0.001137,0.006508,0.005554,0.000759
195774,116646838687_17,116646838687,17,1684863616,FAIL,1,30970295551,361200000000,361217000000,17.0,0.001137,0.006508,0.005554,0.000759
213326,116646838687_17,116646838687,17,606663343078,FAIL,1,32430135524,361200000000,361217000000,17.0,0.001137,0.006508,0.005554,0.000759



Candidate execution: 14423004927_969
Raw rows: 1430
Event types: ['FINISH']
Machines: 1430
Distinct starts: 1
Distinct ends: 1


,_candidate_execution_id,collection_id,instance_index,machine_id,event,failed,time,start_time_us,end_time_us,duration_sec,raw_req_cpu,raw_req_memory,raw_peak_cpu,raw_peak_memory
219237,14423004927_969,14423004927,969,4236312408,FINISH,0,0,1979700000000,1980000000000,300.0,0.001242,0.000202,0.000647,0.000103
37806,14423004927_969,14423004927,969,29952325279,FINISH,0,2322942680,1979700000000,1980000000000,300.0,0.001242,0.000202,0.000647,0.000103
368089,14423004927_969,14423004927,969,144200900,FINISH,0,3282332611,1979700000000,1980000000000,300.0,0.001242,0.000202,0.000647,0.000103
286375,14423004927_969,14423004927,969,2311214894,FINISH,0,4073316663,1979700000000,1980000000000,300.0,0.001242,0.000202,0.000647,0.000103
177971,14423004927_969,14423004927,969,144479369,FINISH,0,6883551076,1979700000000,1980000000000,300.0,0.001242,0.000202,0.000647,0.000103
34248,14423004927_969,14423004927,969,32023110099,FINISH,0,9768014765,1979700000000,1980000000000,300.0,0.001242,0.000202,0.000647,0.000103
103341,14423004927_969,14423004927,969,45859553991,FINISH,0,11682931505,1979700000000,1980000000000,300.0,0.001242,0.000202,0.000647,0.000103
31323,14423004927_969,14423004927,969,4242390821,FINISH,0,13543764567,1979700000000,1980000000000,300.0,0.001242,0.000202,0.000647,0.000103
28914,14423004927_969,14423004927,969,29951828452,FINISH,0,14365546907,1979700000000,1980000000000,300.0,0.001242,0.000202,0.000647,0.000103
261554,14423004927_969,14423004927,969,34204402397,FINISH,0,14888756683,1979700000000,1980000000000,300.0,0.001242,0.000202,0.000647,0.000103



Candidate execution: 122697038507_1191
Raw rows: 1308
Event types: ['ENABLE']
Machines: 1308
Distinct starts: 1
Distinct ends: 1


,_candidate_execution_id,collection_id,instance_index,machine_id,event,failed,time,start_time_us,end_time_us,duration_sec,raw_req_cpu,raw_req_memory,raw_peak_cpu,raw_peak_memory
387472,122697038507_1191,122697038507,1191,549323726,ENABLE,0,1312564766878,1944000000000,1944300000000,300.0,0.013214,0.005211,0.025024,0.003471
250013,122697038507_1191,122697038507,1191,376768600,ENABLE,0,1312708995590,1944000000000,1944300000000,300.0,0.013214,0.005211,0.025024,0.003471
176034,122697038507_1191,122697038507,1191,263445342,ENABLE,0,1313010763933,1944000000000,1944300000000,300.0,0.013214,0.005211,0.025024,0.003471
323803,122697038507_1191,122697038507,1191,263783791,ENABLE,0,1313154709778,1944000000000,1944300000000,300.0,0.013214,0.005211,0.025024,0.003471
266005,122697038507_1191,122697038507,1191,303367904,ENABLE,0,1313735579576,1944000000000,1944300000000,300.0,0.013214,0.005211,0.025024,0.003471
359636,122697038507_1191,122697038507,1191,265322069,ENABLE,0,1313880669610,1944000000000,1944300000000,300.0,0.013214,0.005211,0.025024,0.003471
388690,122697038507_1191,122697038507,1191,1138108493,ENABLE,0,1314310912598,1944000000000,1944300000000,300.0,0.013214,0.005211,0.025024,0.003471
155537,122697038507_1191,122697038507,1191,378446904,ENABLE,0,1314458671933,1944000000000,1944300000000,300.0,0.013214,0.005211,0.025024,0.003471
317395,122697038507_1191,122697038507,1191,799517737688,ENABLE,0,1314603616022,1944000000000,1944300000000,300.0,0.013214,0.005211,0.025024,0.003471
265082,122697038507_1191,122697038507,1191,19535446871,ENABLE,0,1314758012585,1944000000000,1944300000000,300.0,0.013214,0.005211,0.025024,0.003471



Candidate execution: 348020181306_5
Raw rows: 1135
Event types: ['FAIL']
Machines: 1135
Distinct starts: 1
Distinct ends: 1


,_candidate_execution_id,collection_id,instance_index,machine_id,event,failed,time,start_time_us,end_time_us,duration_sec,raw_req_cpu,raw_req_memory,raw_peak_cpu,raw_peak_memory
319873,348020181306_5,348020181306,5,21560261,FAIL,1,618437396,2544777000000,2544778000000,1.0,0.0,0.013016,0.0,9.536743e-07
335837,348020181306_5,348020181306,5,86289635050,FAIL,1,819772230,2544777000000,2544778000000,1.0,0.0,0.013016,0.0,9.536743e-07
360966,348020181306_5,348020181306,5,106885152968,FAIL,1,1762614130,2544777000000,2544778000000,1.0,0.0,0.013016,0.0,9.536743e-07
40846,348020181306_5,348020181306,5,21284212,FAIL,1,2060822600,2544777000000,2544778000000,1.0,0.0,0.013016,0.0,9.536743e-07
337130,348020181306_5,348020181306,5,80667140335,FAIL,1,2368636243,2544777000000,2544778000000,1.0,0.0,0.013016,0.0,9.536743e-07
186338,348020181306_5,348020181306,5,92139577334,FAIL,1,2675884803,2544777000000,2544778000000,1.0,0.0,0.013016,0.0,9.536743e-07
88661,348020181306_5,348020181306,5,23624487592,FAIL,1,2965141844,2544777000000,2544778000000,1.0,0.0,0.013016,0.0,9.536743e-07
169189,348020181306_5,348020181306,5,21375644,FAIL,1,3274880652,2544777000000,2544778000000,1.0,0.0,0.013016,0.0,9.536743e-07
69555,348020181306_5,348020181306,5,265639635884,FAIL,1,3572580236,2544777000000,2544778000000,1.0,0.0,0.013016,0.0,9.536743e-07
22857,348020181306_5,348020181306,5,348056354571,FAIL,1,4417496783,2544777000000,2544778000000,1.0,0.0,0.013016,0.0,9.536743e-07



Candidate execution: 14423004927_1021
Raw rows: 994
Event types: ['FINISH']
Machines: 994
Distinct starts: 1
Distinct ends: 1


,_candidate_execution_id,collection_id,instance_index,machine_id,event,failed,time,start_time_us,end_time_us,duration_sec,raw_req_cpu,raw_req_memory,raw_peak_cpu,raw_peak_memory
369963,14423004927_1021,14423004927,1021,34377925130,FINISH,0,0,1979700000000,1980000000000,300.0,0.001242,0.000202,0.000647,0.000103
80515,14423004927_1021,14423004927,1021,121824482,FINISH,0,4073316935,1979700000000,1980000000000,300.0,0.001242,0.000202,0.000647,0.000103
202529,14423004927_1021,14423004927,1021,34200957193,FINISH,0,5562734902,1979700000000,1980000000000,300.0,0.001242,0.000202,0.000647,0.000103
200088,14423004927_1021,14423004927,1021,155153237,FINISH,0,6883551560,1979700000000,1980000000000,300.0,0.001242,0.000202,0.000647,0.000103
182290,14423004927_1021,14423004927,1021,3302800636,FINISH,0,9768015056,1979700000000,1980000000000,300.0,0.001242,0.000202,0.000647,0.000103
132348,14423004927_1021,14423004927,1021,34254215423,FINISH,0,13604082387,1979700000000,1980000000000,300.0,0.001242,0.000202,0.000647,0.000103
333036,14423004927_1021,14423004927,1021,85631385,FINISH,0,14888756479,1979700000000,1980000000000,300.0,0.001242,0.000202,0.000647,0.000103
268589,14423004927_1021,14423004927,1021,42491983861,FINISH,0,19263369716,1979700000000,1980000000000,300.0,0.001242,0.000214,0.000647,0.000103
134734,14423004927_1021,14423004927,1021,34193965717,FINISH,0,21405020377,1979700000000,1980000000000,300.0,0.001242,0.000214,0.000647,0.000103
18423,14423004927_1021,14423004927,1021,155158551,FINISH,0,22069945447,1979700000000,1980000000000,300.0,0.001242,0.000214,0.000647,0.000103


In [17]:
# =========================================================
# Cell 12: Inspect candidate event/time episodes
# =========================================================
#
# This cell checks whether adding the event and time interval
# separates the broad candidate groups into smaller episodes.
#
# These are diagnostic episode groups only.
# They are not yet the final execution IDs.
# =========================================================

episode_summary = (
    df.groupby(
        [
            "_candidate_execution_id",
            "start_time_us",
            "end_time_us",
            "event",
        ],
        dropna=False,
    )
    .agg(
        raw_row_count=("event", "size"),
        machine_count=("machine_id", "nunique"),
        distinct_time_count=("time", "nunique"),
        failed_row_count=("failed", "sum"),
        peak_cpu_min=("raw_peak_cpu", "min"),
        peak_cpu_max=("raw_peak_cpu", "max"),
        peak_memory_min=("raw_peak_memory", "min"),
        peak_memory_max=("raw_peak_memory", "max"),
    )
    .reset_index()
)

print("Episode groups:", len(episode_summary))

print("\nEpisode-group summary:")
display(
    episode_summary[
        [
            "raw_row_count",
            "machine_count",
            "distinct_time_count",
            "failed_row_count",
        ]
    ].describe()
)


print("\nLargest episode groups:")
display(
    episode_summary
    .sort_values("raw_row_count", ascending=False)
    .head(20)
)


print("\nCandidate groups versus episode groups:")
print(
    "Candidate groups:",
    df["_candidate_execution_id"].nunique()
)
print(
    "Episode groups:",
    len(episode_summary)
)

Episode groups: 253594

Episode-group summary:


,raw_row_count,machine_count,distinct_time_count,failed_row_count
count,253594.000000,253594.000000,253594.000000,253594.000000
mean,1.600566,1.600566,1.599908,0.365458
std,15.606049,15.606049,15.415442,3.340524
min,1.000000,1.000000,1.000000,0.000000
25%,1.000000,1.000000,1.000000,0.000000
50%,1.000000,1.000000,1.000000,0.000000
75%,1.000000,1.000000,1.000000,1.000000
max,6566.000000,6566.000000,6451.000000,1138.000000



Largest episode groups:


,_candidate_execution_id,start_time_us,end_time_us,event,raw_row_count,machine_count,distinct_time_count,failed_row_count,peak_cpu_min,peak_cpu_max,peak_memory_min,peak_memory_max
1947,116646838687_17,2288922000000,2288923000000,ENABLE,6566,6566,6451,0,0.000000,0.000000,9.536743e-07,9.536743e-07
8135,14423004927_969,1979700000000,1980000000000,FINISH,1430,1430,1430,0,0.000647,0.000647,1.029968e-04,1.029968e-04
2155,122697038507_1191,1944000000000,1944300000000,ENABLE,1308,1308,1308,0,0.025024,0.025024,3.471375e-03,3.471375e-03
1946,116646838687_17,361200000000,361217000000,FAIL,1138,1138,1138,1138,0.005554,0.005554,7.591248e-04,7.591248e-04
134845,348020181306_5,2544777000000,2544778000000,FAIL,1135,1135,1135,1135,0.000000,0.000000,9.536743e-07,9.536743e-07
8057,14423004927_1021,1979700000000,1980000000000,FINISH,994,994,994,0,0.000647,0.000647,1.029968e-04,1.029968e-04
53391,260725669813_176,25800000000,26100000000,FINISH,903,903,903,0,0.001392,0.001392,5.294800e-03,5.294800e-03
53271,260725669813_112,25800000000,26100000000,FINISH,758,758,758,0,0.001392,0.001392,5.294800e-03,5.294800e-03
2303,122697038507_5012,1944000000000,1944300000000,ENABLE,686,686,686,0,0.025024,0.025024,3.471375e-03,3.471375e-03
3896,128073966478_2432,115800000000,116041000000,FINISH,586,586,586,0,0.004860,0.004860,1.831055e-04,1.831055e-04



Candidate groups versus episode groups:
Candidate groups: 242946
Episode groups: 253594


In [18]:
# Check whether one Google usage observation is uniquely identified
# by collection, instance, machine, and observation time.

usage_key = [
    "collection_id",
    "instance_index",
    "machine_id",
    "start_time",
    "end_time",
]

key_counts = df.groupby(usage_key, dropna=False).size()

print("Total raw rows:", len(df))
print("Unique observation keys:", key_counts.size)
print("Duplicated observation keys:", int((key_counts > 1).sum()))
print("Maximum rows per observation key:", int(key_counts.max()))

display(key_counts.value_counts().sort_index().head(20))

Total raw rows: 405894
Unique observation keys: 405894
Duplicated observation keys: 0
Maximum rows per observation key: 1


1    405894
Name: count, dtype: int64

## Raw Observation Key

The combination of `collection_id`, `instance_index`, `machine_id`,
`start_time`, and `end_time` uniquely identifies all 405,894 raw rows.

Therefore, the raw dataset contains unique observation records at this
machine-and-time-window grain.

This key is not yet the modeling execution key. Multiple observation records
may still belong to the same reconstructed workload execution.

In [20]:
# Examine the structure of each collection_id + instance_index group.
# This is diagnostic only; it does not create the final execution_id.

candidate_summary = (
    df.groupby(["collection_id", "instance_index"], dropna=False)
      .agg(
          raw_row_count=("machine_id", "size"),
          unique_machine_count=("machine_id", "nunique"),
          unique_start_count=("start_time", "nunique"),
          unique_end_count=("end_time", "nunique"),
          unique_event_count=("event", "nunique"),
          first_start_time=("start_time", "min"),
          last_end_time=("end_time", "max"),
      )
      .reset_index()
)

# Convert the overall time span to seconds for easier interpretation.
candidate_summary["time_span_sec"] = (
    candidate_summary["last_end_time"]
    - candidate_summary["first_start_time"]
) / 1_000_000

# Identify groups where all machines share the same observation window.
candidate_summary["same_observation_window"] = (
    (candidate_summary["unique_start_count"] == 1)
    & (candidate_summary["unique_end_count"] == 1)
)

print("Candidate groups:", len(candidate_summary))

print("\nCandidate groups with multiple machines:")
print(
    (candidate_summary["unique_machine_count"] > 1)
    .value_counts()
)

print("\nCandidate groups with multiple time windows:")
print(
    (
        (candidate_summary["unique_start_count"] > 1)
        | (candidate_summary["unique_end_count"] > 1)
    ).value_counts()
)

print("\nLargest candidate groups:")
display(
    candidate_summary
    .sort_values(
        ["raw_row_count", "unique_machine_count"],
        ascending=False
    )
    .head(20)
)

Candidate groups: 242946

Candidate groups with multiple machines:
unique_machine_count
False    205266
True      37680
Name: count, dtype: int64

Candidate groups with multiple time windows:
False    232646
True      10300
Name: count, dtype: int64

Largest candidate groups:


,collection_id,instance_index,raw_row_count,unique_machine_count,unique_start_count,unique_end_count,unique_event_count,first_start_time,last_end_time,time_span_sec,same_observation_window
36844,116646838687,17,7704,6640,2,2,2,361200000000,2288923000000,1927723.0,False
15314,14423004927,969,1430,1430,1,1,1,1979700000000,1980000000000,300.0,True
37014,122697038507,1191,1308,1308,1,1,1,1944000000000,1944300000000,300.0,True
150117,348020181306,5,1135,1135,1,1,1,2544777000000,2544778000000,1.0,True
15315,14423004927,1021,994,994,1,1,1,1979700000000,1980000000000,300.0,True
73541,260725669813,176,905,903,3,3,3,25800000000,1175400000000,1149600.0,False
73533,260725669813,112,767,758,3,3,3,25800000000,1175400000000,1149600.0,False
37150,122697038507,5012,686,686,1,1,1,1944000000000,1944300000000,300.0,True
38638,128073966478,2432,587,586,2,2,2,115800000000,430500000000,314700.0,False
73535,260725669813,118,580,578,3,3,3,25800000000,1175400000000,1149600.0,False


In [21]:
# Check whether multi-machine groups contain identical resource and event values.
# The grouping excludes machine_id so that all machines in the same
# collection, instance, and observation window are compared together.

window_key = [
    "collection_id",
    "instance_index",
    "start_time",
    "end_time",
]

window_summary = (
    df.groupby(window_key, dropna=False)
      .agg(
          raw_row_count=("machine_id", "size"),
          unique_machine_count=("machine_id", "nunique"),
          unique_event_count=("event", "nunique"),
          unique_req_cpu=("raw_req_cpu", "nunique"),
          unique_req_memory=("raw_req_memory", "nunique"),
          unique_peak_cpu=("raw_peak_cpu", "nunique"),
          unique_peak_memory=("raw_peak_memory", "nunique"),
      )
      .reset_index()
)

multi_machine_windows = window_summary[
    window_summary["unique_machine_count"] > 1
].copy()

print("Multi-machine observation windows:", len(multi_machine_windows))

print("\nResource/event consistency among multi-machine windows:")
display(
    multi_machine_windows[
        [
            "unique_event_count",
            "unique_req_cpu",
            "unique_req_memory",
            "unique_peak_cpu",
            "unique_peak_memory",
        ]
    ]
    .value_counts()
    .head(20)
)

print("\nLargest multi-machine windows:")
display(
    multi_machine_windows
    .sort_values(
        ["unique_machine_count", "raw_row_count"],
        ascending=False
    )
    .head(20)
)

Multi-machine observation windows: 38146

Resource/event consistency among multi-machine windows:


unique_event_count  unique_req_cpu  unique_req_memory  unique_peak_cpu  unique_peak_memory
1                   1               1                  1                1                     19179
                    2               1                  1                1                      7134
                                    2                  1                1                      2663
                    3               1                  1                1                      2548
                    1               2                  1                1                      2228
                    4               1                  1                1                      1227
                    5               1                  1                1                       564
                    2               3                  1                1                       360
                    3               3                  1                1                       322
         


Largest multi-machine windows:


,collection_id,instance_index,start_time,end_time,raw_row_count,unique_machine_count,unique_event_count,unique_req_cpu,unique_req_memory,unique_peak_cpu,unique_peak_memory
38803,116646838687,17,2288922000000,2288923000000,6566,6566,1,2,2,1,1
15991,14423004927,969,1979700000000,1980000000000,1430,1430,1,20,20,1,1
38984,122697038507,1191,1944000000000,1944300000000,1308,1308,1,1,1,1,1
38802,116646838687,17,361200000000,361217000000,1138,1138,1,1,1,1,1
158632,348020181306,5,2544777000000,2544778000000,1135,1135,1,1,1,1,1
15992,14423004927,1021,1979700000000,1980000000000,994,994,1,20,16,1,1
78943,260725669813,176,25800000000,26100000000,903,903,1,22,1,1,1
78925,260725669813,112,25800000000,26100000000,758,758,1,24,1,1,1
39122,122697038507,5012,1944000000000,1944300000000,686,686,1,2,2,1,1
40711,128073966478,2432,115800000000,116041000000,586,586,1,11,38,1,1


In [22]:
# Inspect the largest multi-machine observation window row by row.
# This helps determine whether resource values vary across machines.

largest_window = (
    multi_machine_windows
    .sort_values(
        ["unique_machine_count", "raw_row_count"],
        ascending=False
    )
    .iloc[0]
)

window_mask = (
    (df["collection_id"] == largest_window["collection_id"])
    & (df["instance_index"] == largest_window["instance_index"])
    & (df["start_time"] == largest_window["start_time"])
    & (df["end_time"] == largest_window["end_time"])
)

inspection_cols = [
    "collection_id",
    "instance_index",
    "machine_id",
    "event",
    "failed",
    "start_time",
    "end_time",
    "raw_req_cpu",
    "raw_req_memory",
    "raw_avg_cpu",
    "raw_avg_memory",
    "raw_peak_cpu",
    "raw_peak_memory",
]

print("Rows in selected window:", int(window_mask.sum()))

display(
    df.loc[window_mask, inspection_cols]
      .sort_values(
          ["raw_req_cpu", "raw_req_memory", "machine_id"]
      )
      .head(30)
)

print("\nDistinct values in selected window:")

for col in [
    "event",
    "failed",
    "raw_req_cpu",
    "raw_req_memory",
    "raw_avg_cpu",
    "raw_avg_memory",
    "raw_peak_cpu",
    "raw_peak_memory",
]:
    print(f"\n{col}:")
    print(df.loc[window_mask, col].value_counts(dropna=False).head(10))

Rows in selected window: 6566


,collection_id,instance_index,machine_id,event,failed,start_time,end_time,raw_req_cpu,raw_req_memory,raw_avg_cpu,raw_avg_memory,raw_peak_cpu,raw_peak_memory
295506,116646838687,17,26,ENABLE,0,2288922000000,2288923000000,0.001137,0.006508,0.0,9.536743e-07,0.0,9.536743e-07
212187,116646838687,17,44,ENABLE,0,2288922000000,2288923000000,0.001137,0.006508,0.0,9.536743e-07,0.0,9.536743e-07
236241,116646838687,17,78,ENABLE,0,2288922000000,2288923000000,0.001137,0.006508,0.0,9.536743e-07,0.0,9.536743e-07
197540,116646838687,17,102,ENABLE,0,2288922000000,2288923000000,0.001137,0.006508,0.0,9.536743e-07,0.0,9.536743e-07
401233,116646838687,17,131,ENABLE,0,2288922000000,2288923000000,0.001137,0.006508,0.0,9.536743e-07,0.0,9.536743e-07
80057,116646838687,17,136,ENABLE,0,2288922000000,2288923000000,0.001137,0.006508,0.0,9.536743e-07,0.0,9.536743e-07
114561,116646838687,17,147,ENABLE,0,2288922000000,2288923000000,0.001137,0.006508,0.0,9.536743e-07,0.0,9.536743e-07
326292,116646838687,17,148,ENABLE,0,2288922000000,2288923000000,0.001137,0.006508,0.0,9.536743e-07,0.0,9.536743e-07
306992,116646838687,17,209,ENABLE,0,2288922000000,2288923000000,0.001137,0.006508,0.0,9.536743e-07,0.0,9.536743e-07
117426,116646838687,17,215,ENABLE,0,2288922000000,2288923000000,0.001137,0.006508,0.0,9.536743e-07,0.0,9.536743e-07



Distinct values in selected window:

event:
event
ENABLE    6566
Name: count, dtype: int64

failed:
failed
0    6566
Name: count, dtype: int64

raw_req_cpu:
raw_req_cpu
0.001137    6510
0.009567      56
Name: count, dtype: int64

raw_req_memory:
raw_req_memory
0.006508    6510
0.003784      56
Name: count, dtype: int64

raw_avg_cpu:
raw_avg_cpu
0.0    6566
Name: count, dtype: int64

raw_avg_memory:
raw_avg_memory
9.536743e-07    6566
Name: count, dtype: int64

raw_peak_cpu:
raw_peak_cpu
0.0    6566
Name: count, dtype: int64

raw_peak_memory:
raw_peak_memory
9.536743e-07    6566
Name: count, dtype: int64


In [23]:
# Inspect time and source metadata for the selected multi-machine window.
# This helps determine whether machine rows correspond to distinct
# observation timestamps or repeated source records.

selected = df.loc[window_mask].copy()

window_start = selected["start_time"].iloc[0]
window_end = selected["end_time"].iloc[0]

selected["time_inside_observation_window"] = (
    (selected["time"] >= window_start)
    & (selected["time"] <= window_end)
)

print("Distinct machine IDs:", selected["machine_id"].nunique())
print("Distinct raw time values:", selected["time"].nunique())
print("Minimum raw time:", selected["time"].min())
print("Maximum raw time:", selected["time"].max())
print(
    "Rows with raw time inside observation window:",
    int(selected["time_inside_observation_window"].sum())
)

print("\nRequest/resource combinations:")
display(
    selected.groupby(
        [
            "raw_req_cpu",
            "raw_req_memory",
            "raw_peak_cpu",
            "raw_peak_memory",
            "event",
            "failed",
        ],
        dropna=False
    )
    .size()
    .reset_index(name="row_count")
    .sort_values("row_count", ascending=False)
)

print("\nMetadata consistency:")
metadata_cols = [
    "collection_type",
    "priority",
    "alloc_collection_id",
    "user",
    "collection_name",
    "collection_logical_name",
    "scheduler",
    "vertical_scaling",
]

display(
    selected[metadata_cols]
    .nunique(dropna=False)
    .sort_values(ascending=False)
)

print("\nSample rows ordered by raw time:")
display(
    selected.sort_values("time")[
        [
            "time",
            "machine_id",
            "event",
            "raw_req_cpu",
            "raw_req_memory",
            "raw_peak_cpu",
            "raw_peak_memory",
        ]
    ].head(30)
)

Distinct machine IDs: 6566
Distinct raw time values: 6451
Minimum raw time: 0
Maximum raw time: 2678923967375
Rows with raw time inside observation window: 0

Request/resource combinations:


,raw_req_cpu,raw_req_memory,raw_peak_cpu,raw_peak_memory,event,failed,row_count
0,0.001137,0.006508,0.0,9.536743e-07,ENABLE,0,6510
1,0.009567,0.003784,0.0,9.536743e-07,ENABLE,0,56



Metadata consistency:


collection_type            1
priority                   1
alloc_collection_id        1
user                       1
collection_name            1
collection_logical_name    1
scheduler                  1
vertical_scaling           1
dtype: int64


Sample rows ordered by raw time:


,time,machine_id,event,raw_req_cpu,raw_req_memory,raw_peak_cpu,raw_peak_memory
369091,0,67376881683,ENABLE,0.009567,0.003784,0.0,9.536743e-07
165158,0,30633179879,ENABLE,0.009567,0.003784,0.0,9.536743e-07
135455,0,667262623721,ENABLE,0.009567,0.003784,0.0,9.536743e-07
308824,0,175802551009,ENABLE,0.001137,0.006508,0.0,9.536743e-07
11074,0,175787406670,ENABLE,0.001137,0.006508,0.0,9.536743e-07
190813,0,61194791580,ENABLE,0.009567,0.003784,0.0,9.536743e-07
86486,0,291385060,ENABLE,0.009567,0.003784,0.0,9.536743e-07
134903,0,175831796786,ENABLE,0.001137,0.006508,0.0,9.536743e-07
362497,0,557670727557,ENABLE,0.009567,0.003784,0.0,9.536743e-07
330660,0,61377165770,ENABLE,0.001137,0.006508,0.0,9.536743e-07


## Time and Event-Usage Alignment Issue

For a selected multi-machine group, all rows have `time = 0`, while
`start_time` and `end_time` define a later usage window.

The rows also contain thousands of machine IDs with identical event and
peak-usage values. This indicates that the CSV may combine event-level and
usage-level fields at different grains.

The raw `time` field must not be used for execution ordering until its
relationship with `start_time`, `end_time`, and event type is verified.

In [24]:
# Compare the raw event time with the usage observation window.
# This checks whether time is aligned with start_time/end_time
# for each event type.

time_alignment = (
    df.groupby("event", dropna=False)
      .agg(
          row_count=("event", "size"),
          zero_time_rows=("time", lambda x: (x == 0).sum()),
          distinct_time_values=("time", "nunique"),
          distinct_start_values=("start_time", "nunique"),
          distinct_end_values=("end_time", "nunique"),
      )
      .reset_index()
)

time_alignment["zero_time_percentage"] = (
    time_alignment["zero_time_rows"]
    / time_alignment["row_count"]
    * 100
)

display(
    time_alignment.sort_values(
        "row_count",
        ascending=False
    )
)

# Check whether raw time falls inside the usage window.
df["time_inside_usage_window"] = (
    (df["time"] >= df["start_time"])
    & (df["time"] <= df["end_time"])
)

print(
    "Rows with time inside usage window:",
    int(df["time_inside_usage_window"].sum()),
    "of",
    len(df)
)

print(
    "Percentage:",
    f"{df['time_inside_usage_window'].mean() * 100:.2f}%"
)

,event,row_count,zero_time_rows,distinct_time_values,distinct_start_values,distinct_end_values,zero_time_percentage
3,FINISH,92867,600,92199,909,911,0.646085
2,FAIL,92678,37082,55576,1271,1269,40.011653
0,ENABLE,75907,18616,57288,981,978,24.524747
7,SCHEDULE,69104,154,68709,293,293,0.222853
5,LOST,59515,0,59515,268,267,0.000000
1,EVICT,14756,0,84,24,24,0.000000
4,KILL,951,0,951,895,877,0.000000
8,UPDATE_PENDING,111,0,44,44,45,0.000000
6,QUEUE,4,0,4,2,2,0.000000
9,UPDATE_RUNNING,1,0,1,1,1,0.000000


Rows with time inside usage window: 1506 of 405894
Percentage: 0.37%
